File: languages/perl/tests/position_tape_test.pl
use strict;
use warnings;
use Digest::SHA qw(sha256_hex);
use FindBin qw($Bin);
use lib "$Bin/../src";
use PositionTape qw(
  Generate GenerateMarkerComplete Locate Validate FindTruncationPoint
  FindFirstMismatch BuildWindowIndex LocateByHash HashFragment
);

sub assert_equal {
    my ($expected, $actual, $message) = @_;
    die "$message: got <$actual>, want <$expected>\n" if !defined($expected) || !defined($actual) || $expected ne $actual;
}

assert_equal("", Generate(0), "Generate(0)");
assert_equal("1234567891", Generate(10), "Generate(10)");
assert_equal(101, length(GenerateMarkerComplete(100)), "GenerateMarkerComplete(100)");
assert_equal(10003, length(GenerateMarkerComplete(10000)), "GenerateMarkerComplete(10000)");

my $expected = Generate(50);
assert_equal(1, Validate($expected, 50)->{is_valid}, "valid result");
assert_equal(18, Validate(substr($expected, 0, 17), 50)->{truncation_point}, "truncation point");
assert_equal(4, FindTruncationPoint("123X"), "mismatch point");
assert_equal(13, FindFirstMismatch($expected, substr($expected, 0, 12) . "X" . substr($expected, 13))->{position}, "mismatch");

my $fragment = substr(Generate(80), 29, 12);
my $hash = HashFragment($fragment);
assert_equal(30, Locate($fragment), "Locate");
die "BuildWindowIndex missing position 30\n" unless grep { $_ == 30 } @{ BuildWindowIndex(length($fragment))->{$hash} };
die "LocateByHash missing position 30\n" unless grep { $_ == 30 } @{ LocateByHash(uc($hash), length($fragment)) };

open my $manifest_fh, "<:raw", "$Bin/../../../fixtures/manifest.generated.json" or die $!;
my $manifest = do { local $/; <$manifest_fh> };
while ($manifest =~ /"file":\s*"([^"]+)".*?"bytes":\s*(\d+).*?"sha256":\s*"([^"]+)"/gs) {
    my ($file, $bytes, $sha) = ($1, $2, $3);
    open my $fixture_fh, "<:raw", "$Bin/../../../fixtures/$file" or die $!;
    my $raw = do { local $/; <$fixture_fh> };
    assert_equal($bytes, length($raw), "$file bytes");
    assert_equal($sha, sha256_hex($raw), "$file sha256");
    die "$file has UTF-8 BOM\n" if substr($raw, 0, 3) eq "\xef\xbb\xbf";
    die "$file has trailing newline\n" if $raw =~ /[\r\n]\z/;
}

print "OK perl position_tape\n";

File: languages/ruby/tests/position_tape_test.rb
require "digest"
require "json"
require_relative "../src/position_tape"

def assert_equal(expected, actual, message)
  raise "#{message}: got #{actual.inspect}, want #{expected.inspect}" unless expected == actual
end

assert_equal("", PositionTape.Generate(0), "Generate(0)")
assert_equal("1234567891", PositionTape.Generate(10), "Generate(10)")
assert_equal(101, PositionTape.GenerateMarkerComplete(100).length, "GenerateMarkerComplete(100)")
assert_equal(10_003, PositionTape.GenerateMarkerComplete(10_000).length, "GenerateMarkerComplete(10000)")

expected = PositionTape.Generate(50)
assert_equal(true, PositionTape.Validate(expected, 50).is_valid, "valid result")
assert_equal(18, PositionTape.Validate(expected[0, 17], 50).truncation_point, "truncation point")
assert_equal(4, PositionTape.FindTruncationPoint("123X"), "mismatch point")
assert_equal(13, PositionTape.FindFirstMismatch(expected, "#{expected[0, 12]}X#{expected[13..]}").position, "mismatch")

fragment = PositionTape.Generate(80)[29, 12]
assert_equal("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", PositionTape.HashFragment(""), "sha256 empty")
assert_equal("ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad", PositionTape.HashFragment("abc"), "sha256 abc")
hash = PositionTape.HashFragment(fragment)
assert_equal(30, PositionTape.Locate(fragment), "Locate")
raise "BuildWindowIndex missing position 30" unless PositionTape.BuildWindowIndex(fragment.length)[hash].include?(30)
raise "LocateByHash missing position 30" unless PositionTape.LocateByHash(hash.upcase, fragment.length).include?(30)

root = File.expand_path("../../..", __dir__)
manifest = JSON.parse(File.read(File.join(root, "fixtures", "manifest.generated.json")))
manifest["fixtures"].each do |fixture|
  raw = File.binread(File.join(root, "fixtures", fixture["file"]))
  assert_equal(fixture["bytes"], raw.bytesize, "#{fixture['file']} bytes")
  assert_equal(fixture["sha256"], Digest::SHA256.hexdigest(raw), "#{fixture['file']} sha256")
  raise "#{fixture['file']} has UTF-8 BOM" if raw.start_with?("\xef\xbb\xbf".b)
  raise "#{fixture['file']} has trailing newline" if raw.end_with?("\n") || raw.end_with?("\r")
end

puts "OK ruby position_tape"
